eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,404 @@
+				}
+				// On success, terminate the loop
+				return
+			}
+		}
+	}()
+
+	return logs
+}
+
+// Healthy returns if execution client is currently healthy: responds to requests and not in the syncing state.
+func (ec *MultiClient) Healthy(ctx context.Context) error {
+	f := func(client SingleClientProvider) (any, error) {
+		return nil, client.Healthy(ctx)
+	}
+
+	if _, err := ec.call(ec.setMethod(ctx, "Healthy"), f); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// BlockByNumber retrieves a block by its number.
+func (ec *MultiClient) BlockByNumber(ctx context.Context, blockNumber *big.Int) (*ethtypes.Block, error) {
+	f := func(client SingleClientProvider) (any, error) {
+		return client.BlockByNumber(ctx, blockNumber)
+	}
+	res, err := ec.call(ec.setMethod(ctx, "BlockByNumber"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.(*ethtypes.Block), nil
+}
+
+// HeaderByNumber retrieves a block header by its number.
+func (ec *MultiClient) HeaderByNumber(ctx context.Context, blockNumber *big.Int) (*ethtypes.Header, error) {
+	f := func(client SingleClientProvider) (any, error) {
+		return client.HeaderByNumber(ctx, blockNumber)
+	}
+	res, err := ec.call(ec.setMethod(ctx, "HeaderByNumber"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.(*ethtypes.Header), nil
+}
+
+func (ec *MultiClient) SubscribeFilterLogs(ctx context.Context, q ethereum.FilterQuery, ch chan<- ethtypes.Log) (ethereum.Subscription, error) {
+	f := func(client SingleClientProvider) (any, error) {
+		return client.SubscribeFilterLogs(ctx, q, ch)
+	}
+	res, err := ec.call(ec.setMethod(ctx, "SubscribeFilterLogs"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.(ethereum.Subscription), nil
+}
+
+func (ec *MultiClient) FilterLogs(ctx context.Context, q ethereum.FilterQuery) ([]ethtypes.Log, error) {
+	f := func(client SingleClientProvider) (any, error) {
+		return client.FilterLogs(ctx, q)
+	}
+	res, err := ec.call(ec.setMethod(ctx, "FilterLogs"), f)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.([]ethtypes.Log), nil
+}
+
+func (ec *MultiClient) Filterer() (*contract.ContractFilterer, error) {
+	return contract.NewContractFilterer(ec.contractAddress, ec)
+}
+
+func (ec *MultiClient) ChainID(_ context.Context) (*big.Int, error) {
+	return ec.chainID, nil
+}
+
+func (ec *MultiClient) Close() error {
+	close(ec.closed)
+
+	var multiErr error
+	for i, client := range ec.clients {
+		if err := client.Close(); err != nil {
+			ec.logger.Debug("Failed to close client", zap.String("address", ec.nodeAddrs[i]), zap.Error(err))
+			multiErr = errors.Join(multiErr, err)
+		}
+	}
+
+	return multiErr
+}
+
+func (ec *MultiClient) call(ctx context.Context, f func(client SingleClientProvider) (any, error)) (any, error) {
+	if len(ec.clients) == 1 {
+		return f(ec.clients[0])
+	}
+
+	for i := 0; i < len(ec.clients); i++ {
+		ec.currClientMu.Lock()
+		currentIdx := ec.currClientIdx
+		client := ec.clients[currentIdx]
```

Comment:
is `clients` ever changing ? I don't think there's a reason to lock it.

Reply:
Only `currClientIdx` changes. Indeed, this line can be moved outside the lock